src/cose/mac0.rs

```diff
@@ -0,0 +1,546 @@
+use std::collections::BTreeMap;
+
+use ::hmac::Hmac;
+use coset::cbor::Value;
+use coset::cwt::ClaimsSet;
+use coset::{
+    iana, mac_structure_data, AsCborValue, CborSerializable, CoseError, MacContext,
+    RegisteredLabelWithPrivate, TaggedCborSerializable,
+};
+use digest::{Mac, MacError};
+use isomdl_macros::FieldsNames;
+use sha2::Sha256;
+
+use crate::cbor::CborValue;
+use crate::cose::SignatureAlgorithm;
+
+/// Prepared `COSE_Mac0` for remote signing.
+///
+/// To produce a `COSE_Mac0` do the following:
+///
+/// 1. Set the signature algorithm with [coset::HeaderBuilder::algorithm].
+/// 2. Produce a signature remotely, according to the chosen signature algorithm,
+///    using the [PreparedCoseMac0::signature_payload] as the payload.
+/// 3. Generate the `COSE_Mac0` by passing the produced signature into
+///    [PreparedCoseMac0::finalize].
+///
+/// Example:
+/// ```
+/// use coset::iana;
+/// use digest::Mac;
+/// use hex::FromHex;use hmac::Hmac;
+/// use sha2::Sha256;
+/// use isomdl::cose::mac0::PreparedCoseMac0;
+///
+/// let key = Vec::<u8>::from_hex("a361316953796d6d6574726963613305622d318f187418681869187318201869187318201874186818651820186b18651879").unwrap();
+/// let signer = Hmac::<Sha256>::new_from_slice(&key).expect("failed to create HMAC signer");
+/// let protected = coset::HeaderBuilder::new()
+///     .algorithm(iana::Algorithm::HMAC_256_256)
+///     .build();
+/// let unprotected = coset::HeaderBuilder::new().key_id(b"11".to_vec()).build();
+/// let builder = coset::CoseMac0Builder::new()
+///     .protected(protected)
+///     .unprotected(unprotected)
+///     .payload(b"This is the content.".to_vec());
+/// let prepared = PreparedCoseMac0::new(builder, None, None, true).unwrap();
+/// let signature_payload = prepared.signature_payload();
+/// let signature = tag(signature_payload, &signer).unwrap();
+/// let cose_mac0 = prepared.finalize(signature);
+/// fn tag(signature_payload: &[u8], s: &Hmac<Sha256>) -> anyhow::Result<Vec<u8>> {
+///     let mut mac = s.clone();
+///     mac.reset();
+///     mac.update(signature_payload);
+///     Ok(mac.finalize().into_bytes().to_vec())
+///  }
+#[derive(Debug, Clone, FieldsNames)]
+pub struct PreparedCoseMac0 {
+    tagged: bool,
```

Comment:
Isn't this redundant with `cose_mac0.tagged`?

Reply:
added this so we can propagate to `CoseMac0`.

Reply:
Added `MaybeTagged` and have `tagged` just in there.